embeddings/hyperparameter_search/configspace.py

```diff
@@ -0,0 +1,191 @@
+import abc
+from abc import ABC
+from dataclasses import dataclass, field
+from typing import Any, Dict, Final, Tuple, Type, TypeVar, Union
+
+import optuna
+
+from embeddings.hyperparameter_search.parameters import ConstantParameter, SearchableParameter
+from embeddings.utils.utils import PrimitiveTypes
+
+Parameter = Union[SearchableParameter, ConstantParameter]
+ParsedParameters = TypeVar("ParsedParameters")
+SampledParameters = Dict[str, Union[PrimitiveTypes, Dict[str, PrimitiveTypes]]]
+
+
+class ConfigSpace(ABC):
+    def _parse_parameter(
+        self, param_name: str, trial: optuna.trial.Trial
+    ) -> Tuple[str, PrimitiveTypes]:
+        param: Parameter = self.__getattribute__(param_name)
+        if isinstance(param, SearchableParameter):
+            return param.name, trial._suggest(name=param.name, distribution=param.distribution)
+        elif isinstance(param, ConstantParameter):
+            return param.name, param.value
+        else:
+            raise ValueError(
+                f"Parameter type {type(param)} is not suported! "
+                "Supported types are: SearchableParameter and ConstantParameter"
+            )
+
+    def _map_parameters(
+        self, trial: optuna.trial.Trial, parameters_types: Dict[str, Any]
+    ) -> Dict[str, PrimitiveTypes]:
+        parameters: Dict[str, PrimitiveTypes] = {}
+        for param_name in parameters_types.keys():
+            parameters.update([self._parse_parameter(trial=trial, param_name=param_name)])
+
+        return parameters
+
+    @classmethod
+    def _get_annotations(cls) -> Dict[str, Type[Parameter]]:
+        annotations = {}
+        for c in cls.mro():
+            annotations.update(getattr(c, "__annotations__", {}))
+        return annotations
+
+    def sample_parameters(self, trial: optuna.trial.Trial) -> Dict[str, PrimitiveTypes]:
+        return self._map_parameters(
+            parameters_types={
+                param_name: param_cfg for param_name, param_cfg in self._get_annotations().items()
```

Comment:
if I'm correct we only need parameter names right? we could retrieve only the names without the config then

Reply:
Yes you are correct. I will handle it in next refactor